src/storage/src/decode/mod.rs

```diff
@@ -373,88 +393,105 @@ where
-                            partition: partition.clone(),
-                            metadata: to_metadata_row(
-                                &metadata_items,
-                                partition.clone(),
-                                *position,
-                                *upstream_time_millis,
-                                headers.as_deref(),
-                            ),
-                        };
-                        session.give((result, ts.clone(), *diff));
-                    }
-                });
-                // Matching historical practice, we only log metrics on the value decoder.
-                if n_errors > 0 {
-                    value_decoder.log_errors(n_errors);
-                }
-                if n_successes > 0 {
-                    value_decoder.log_successes(n_successes);
+            let mut n_errors = 0;
+            let mut n_successes = 0;
+            for (output, ts, diff) in data.iter() {
+                let SourceOutput {
+                    key,
+                    value,
+                    position,
+                    upstream_time_millis,
+                    partition,
+                    headers,
+                } = output;
+
+                let key = match key_decoder.as_mut().zip(key.as_ref()) {
+                    Some((decoder, buf)) => decode_delimited(decoder, buf).await.transpose(),
+                    None => None,
+                };
+
+                let value = match value.as_ref() {
+                    Some(buf) => decode_delimited(&mut value_decoder, buf).await.transpose(),
+                    None => None,
+                };
+
+                if matches!(&key, Some(Err(_))) || matches!(&value, Some(Err(_))) {
+                    n_errors += 1;
+                } else if matches!(&value, Some(Ok(_))) {
+                    n_successes += 1;
                 }
+
+                let result = DecodeResult {
+                    key,
+                    value,
+                    position: *position,
+                    upstream_time_millis: *upstream_time_millis,
+                    partition: partition.clone(),
+                    metadata: to_metadata_row(
+                        &metadata_items,
+                        partition.clone(),
+                        *position,
+                        *upstream_time_millis,
+                        headers.as_deref(),
+                    ),
+                };
+                output_container.push((result, ts.clone(), *diff));
+            }
+
+            // Matching historical practice, we only log metrics on the value decoder.
+            if n_errors > 0 {
+                value_decoder.log_errors(n_errors);
+            }
+            if n_successes > 0 {
+                value_decoder.log_successes(n_successes);
             }
-        });
-    (results.as_collection(), None)
+
+            output_handle
+                .give_container(&cap, &mut output_container)
```

Comment:
In defense of the classical operators, they can also express this pattern but we usually don't need to because the session automatically batch things together. The `handle.give()` function of async handles also batches things together but it incurs the additional cost of checking if we should yield and if the capability is valid for the given output for every item.

I should eventually come to it and bring back well functioning sessions to async operators